src/assignment-5/CollateralizedVault.sol

```diff
@@ -0,0 +1,168 @@
+        }
+
+        borrows[msg.sender] += amount;
+
+        underlying.transfer(msg.sender, amount);
+    }
+
+    /// @notice Pays back an open debt
+    /// @param amount The amount of `underlying` token to pay back
+    function repay(uint256 amount) public {
+        borrows[msg.sender] -= amount;
+
+        underlying.transferFrom(msg.sender, address(this), amount);
+    }
+
+    /// @notice Withdraws part of the collateral
+    ///     Only allowed to withdraw as long as the collateral left is higher in value than the debt
+    /// @param collateralAmount The amount of `collateral` token to withdraw
+    function withdraw(uint256 collateralAmount) public {
+        uint256 requiredCollateral = getRequiredCollateral(borrows[msg.sender]);
+
+        if (deposits[msg.sender] - collateralAmount < requiredCollateral) {
+            revert TooMuchDebt();
+        }
+
+        deposits[msg.sender] -= collateralAmount;
+
+        collateral.transfer(msg.sender, collateralAmount);
+    }
+
+    /// @notice Admin: liquidate a user debt if the collateral value falls below the debt
+    /// @param user The user to liquidate
+    /// @dev Only admin is allowed to liquidate
+    function liquidate(address user) onlyOwner public {
+        uint256 requiredCollateral = getRequiredCollateral(borrows[user]);
+        if (deposits[user] >= requiredCollateral) {
+            revert UserDebtIsSufficientlyCollateralized();
+        }
+
+        delete borrows[user];
+        delete deposits[user];
+    }
+
+    /// @notice Returns the required amount of collateral in order to borrow `borrowAmount`
+    function getRequiredCollateral(uint256 borrowAmount) public view returns (uint256 requiredCollateral) {
+        requiredCollateral = mulup(borrowAmount, getPrice(), oracleDecimals);
+        requiredCollateral = scaleInteger(requiredCollateral, underlyingDecimals, collateralDecimals);
+    }
+
+    /// @dev Scales a fixed point interger from `fromDecimals` to `toDecimals`
+    function scaleInteger(uint256 x, uint8 fromDecimals, uint8 toDecimals) public pure returns (uint256) {
+        if (toDecimals >= fromDecimals) {
+            return x * 10**(toDecimals - fromDecimals);
+        } else {
+            return x / 10**(fromDecimals - toDecimals);
+        }
+    }
+
+    /// @notice Returns the price of the underlying token in collateral token units
+    function getPrice() public view returns (uint) {
+        (
+            /*uint80 roundID*/,
+            int price,
+            /*uint startedAt*/,
+            /*uint timeStamp*/,
+            /*uint80 answeredInRound*/
+        ) = oracle.latestRoundData();
+        return toUint256(price);
+    }
+
+    function mulup(uint256 x, uint256 y, uint8 decimals) internal pure returns (uint256 z) {
+        z = (x * y + 10**decimals - 1) / 10**decimals;
+    }
+
+    function toUint256(int256 value) internal pure returns (uint256) {
+        require(value >= 0, "SafeCast: value must be positive");
```

Comment:
this doesn't need to be changed, but in case ur not aware both this is and the fn above are better suited for  libraries.  

that way u can use the `using CastI256U256 for uint256;` [syntax](https://docs.soliditylang.org/en/v0.8.13/contracts.html?highlight=libraries%20as#libraries)? 

if this were for production u would add ur own CastI256U256 lib here https://github.com/yieldprotocol/yield-utils-v2/tree/main/contracts/cast

Reply:
💯